src/abi-provider.ts

```diff
@@ -0,0 +1,196 @@
+import fs from "node:fs";
+import path from "node:path";
+
+import { confirm as askUserToConfirm } from "@inquirer/prompts";
+import chalk from "chalk";
+import jsonDiff from "json-diff";
+
+import { printError } from "./common";
+import { log, LogCommand, logError, logErrorAndExit } from "./logger";
+import { g_Arguments } from "./state-mate";
+import { Abi, ContractInfo, isValidAbi } from "./types";
+
+function loadAbiFromAbiPath(abiPath: string): Abi | never {
+  try {
+    const abiFileContent = fs.readFileSync(abiPath, "utf8");
+    const abiJson: unknown = JSON.parse(abiFileContent);
+
+    const abi: unknown = abiJson instanceof Object && "abi" in abiJson ? abiJson.abi : abiJson;
+    if (!isValidAbi(abi)) {
+      logErrorAndExit(`ABI file ${abiPath} does not contain valid ABI`);
+    }
+    return abi;
+  } catch (error) {
+    logErrorAndExit(`Failed to read ABI file at ${abiPath}: ${printError(error)} `);
+  }
+}
+
+export function loadAbiFromFile(contractName: string, address: string): Abi | never {
+  address = address.toLowerCase();
+  let abiPath;
+
+  try {
+    abiPath = _findAbiPath(contractName, address, { shouldThrow: true });
+  } catch (error) {
+    logErrorAndExit(
+      `Error finding ABI file for contract
+        ${contractName} in ${g_Arguments.abiDirPath}: ${printError(error)}\n\n` +
+        chalk.yellow.bold(`Try running with the 'abi' option to download the unnecessary ABI`),
+    );
+  }
+  return loadAbiFromAbiPath(abiPath);
+}
+
+export async function checkAllAbi(contractInfo: ContractInfo) {
+  const { contractName, address, abi, implementation } = contractInfo;
+  await _checkAbi(contractName, address, abi);
+  if (implementation) {
+    await _checkAbi(implementation.contractName, address, implementation.abi);
+    await _checkAbi(implementation.contractName, implementation.address, implementation.abi);
+  }
+}
+
+async function _checkAbi(contractName: string, address: string, abiFromExplorer: Abi): Promise<void> {
+  address = address.toLowerCase();
+  const logHandler = new LogCommand(`ABI ${chalk.magenta(`${contractName}-${address}.json`)}`);
+  let isNeededToSaveAbi = true;
+  const abiExistedPath = _findAbiPath(contractName, address, { shouldThrow: false });
+
+  if (abiExistedPath) {
+    isNeededToSaveAbi = await _isNeededToOverwriteExistedAbi(abiExistedPath, abiFromExplorer);
+  }
+
+  if (isNeededToSaveAbi) {
+    const abiFileNameToSave = abiExistedPath || _defaultAbiFilePath(contractName, address);
+    _saveAbi(abiFileNameToSave, abiFromExplorer);
+    logHandler.success(abiExistedPath ? "Overwritten" : "Saved");
+  }
+  async function _isNeededToOverwriteExistedAbi(abiExistedPath: string, abiFromExplorer: Abi): Promise<boolean> {
```

Comment:
can simplify this a lot (basically get rid of it) if we take the unconditional `update-abi` road